### DISCUSSION TITLE
Looking for a reference implementation for FCM token handling

### DISCUSSION THREAD

**Discussion by Yonom** [invertase/react-native-firebase#5904] [Q&A]
The ["Best practices for FCM registration token management"](https://firebase.google.com/docs/cloud-messaging/manage-tokens) document suggests several things to implement:

- storing and updating registration tokens along with a timestamp
- background service in app to update the tokens every month
- locally storing the list of topic subscriptions
- resubscribing to all topic subscriptions every month & whenever the token changes
- removing stale tokens from subscriptions if the token is older than 2 months

Is there a reference implementation or some open source project that implements all this? (e. g. using firestore as the backend)

**Comment by andersonaddo** [comment]
You'll be hard pressed to find a codebase online that does all of that. Fcm token handling is very project specific, and some of these suggestions certainly aren't needed for every project.

I recently open-sourced my closed down startup, and I think we had some pretty clean and scalable code for fcm token handling. But, before you look at it, like i said, it doesn't do everything in this list since not everything is necessary. So, we don't update tokens every month: we update them on the backend every time there's a token change locally (which can happen more often than monthly). We also don't really use topic subscriptions: only tokens were good enough for us. Lastly, we don't have a ttl for tokens to remove them after 2 months. Rather, we remove all tokens that can no longer be used to send notifications (which means _something_ (but not _necessarily_ age) has made that token invalid).

Here's the code for a lot of the core fcm handling functions on the backend 
https://github.com/emitapp/emit-server-public/blob/master/functions/src/fcmFunctions/fcmCore.ts

And heres where a lot of the client side fcm handling is done (executed on every fresh launch of the app, as long as the user is logged in)
https://github.com/emitapp/emit-public/blob/master/src/screens/MainTabNav.js#L173

There's a lot of app-specific things done on the client side, but I'd actually say the server side is pretty clean so it should be easy to understnad. I also took the pain to make sure this part of my project is well commented since fcm is not fun to re-learn.

**Reply by mikehardy** [reply]
It is never fantastic to hear of a closed down startup @andersonaddo but that is really cool of you to open up the code.
To the question: I haven't seen any reference codebases either, and in my main work project where we use messaging, we're doing just enough to get by - that is, we are also not crawling all tokens and refreshing. I maintain an FCM token map from the installations id to the token in each user document in my users collection in firestore, and I use a function to send to that user, if any token comes back as unusable I remove it from the map. If the user for some reason gets a new installation id, I add it to the map with current token, any time token changes on same installation id I update it. That's about it
